### PR TITLE
frame/result: allow differing TableSpecs in PreparedMetadata

### PR DESCRIPTION
**Pull request by wprzytula**
As noted in #1134, when preparing batches containing requests to multiple to different tables, the PreparedMetadata received upon preparation contains differing TableSpecs (i.e., TableSpecs with more than table mentioned).
This fails the check that we introduced with ResultMetadata in mind: we've been assuming that all TableSpecs are the same, because ScyllaDB/ Cassandra has no support for JOINs.
Not to fail preparation of cross-table batches, the check is now only turned on for ResultMetadata and turned off for PreparedMetadata.

Fixes: #1134

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

### CODE REVIEW

**Review by muzarski — COMMENTED** [review]
Looks good, but can we add a regression test for the scenario mentioned in issue (i.e., where we prepare a batch statement with bind markers)?

**Comment by wprzytula** [comment]
Rebased on main.

File: scylla/src/statement/batch.rs [hunk 1]
```diff
@@ -178,6 +178,12 @@ impl From<&str> for BatchStatement {
     }
 }
 
+impl From<String> for BatchStatement {
+    fn from(s: String) -> Self {
+        BatchStatement::Query(Query::from(s))
+    }
+}
+
```
Comment (muzarski) [inline scylla/src/statement/batch.rs:186]: Was this change required to implement the test (it's in the same commit)?

File: scylla/src/transport/session_test.rs [hunk 2]
```diff
@@ -537,6 +537,42 @@ async fn test_batch() {
     assert_eq!(results, vec![(4, 20, String::from("foobar"))]);
 }
 
+// This is a regression test for #1134.
+#[tokio::test]
+async fn test_batch_to_multiple_tables() {
+    setup_tracing();
+    let session = Arc::new(create_new_session_builder().build().await.unwrap());
```
Comment (muzarski) [inline scylla/src/transport/session_test.rs:544]: nit: is Arc required here?

**Review by muzarski — APPROVED** [review]

File: scylla/src/statement/batch.rs (hunk 1 above)
Comment (wprzytula) [reply scylla/src/statement/batch.rs:186]: Good catch. I included this by accident, after I noticed such implementation is missing.

File: scylla/src/transport/session_test.rs (hunk 2 above)
Comment (wprzytula) [reply scylla/src/transport/session_test.rs:544]: Not at all, indeed.

File: scylla/src/statement/batch.rs (hunk 1 above)
Comment (wprzytula) [reply scylla/src/statement/batch.rs:186]: I'm move this implementation out of this PR.

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Now that the metadata supports zero-copy deserialization, what is the benefit of having this check at all? If the only benefit is that some objects are a bit smaller, then this is imo not enough to justify additional complexity.
Diverging from CQL protocol have just hurt us, so we should avoid letting it hurt us again in the future.

**Comment by wprzytula** [comment]
> Now that the metadata supports zero-copy deserialization, what is the benefit of having this check at all? If the only benefit is that some objects are a bit smaller, then this is imo not enough to justify additional complexity. Diverging from CQL protocol have just hurt us, so we should avoid letting it hurt us again in the future.

OK, I'll remove the check at all.

**Comment by Lorak-mmk** [comment]
Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?

**Comment by wprzytula** [comment]
> Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?

What is the "it" that you are referring to? What do you consider to be a performance optimisation? Because the check itself is for sure not an optimisation.

**Comment by wprzytula** [comment]
v2:
- addressed @muzarski's comments,
- deleted the check (and the corresponding error variant) completely. All things considered, it's no use having it, especially once we established that we are not going to decouple TableSpec from ColumnSpec. We also no longer have strong motivation for it, because the ResultMetadata is now always deserialized in a borrowed form, which is much more lightweight.

A step that should actually be performed in that regard is pushing https://github.com/scylladb/scylladb/issues/17788 forwards, because the current way it's done is a dramatic waste.

**Comment by Lorak-mmk** [comment]
> > Ok. If there are other considerations here (e.g. simpler API) then it may be worth keeping the check. So the question remains - is it only a performance opt or something more?
> 
> What is the "it" that you are referring to? What do you consider to be a performance optimisation? Because the check itself is for sure not an optimisation.

Sorry, I was writing the comment quickly and hastily.
The thing is I don't remember all the discussion and API changes around this.

What I meant is that if assuming the same table spec for columns gives us some API advantages, like being able to expose a table spec directly on query response (compared to expising it on the column metadata of the response), then it may be worth considering.

Getting rid of the assumption also lets us get rid of some types (like ColumnSpecView / ColumnSpecs), right? Then imo it is argument for that because it removes and simplifies code.

**Comment by wprzytula** [comment]
What I like about `ColumnSpecs` is that it contains logic for `get_by_name` retrieval, which is absent on plain `&[ColumnSpec]`. I'd like to keep `ColumnSpecs` as such in our codebase. We could indeed remove `TableSpecView` and `ColumnSpecView`, and also provide an `inner` method for accessing `&[ColumnSpec]` from `ColumnSpecs`.

**Comment by Lorak-mmk** [comment]
> What I like about `ColumnSpecs` is that it contains logic for `get_by_name` retrieval, which is absent on plain `&[ColumnSpec]`. I'd like to keep `ColumnSpecs` as such in our codebase. We could indeed remove `TableSpecView` and `ColumnSpecView`, and also provide an `inner` method for accessing `&[ColumnSpec]` from `ColumnSpecs`.
> 
> cc @muzarski

Ok, makes sense.

**Comment by wprzytula** [comment]
This is API-breaking due to an error variant being removed. I suggest leaving the error variant for now under a TODO and removing it in the next major release. cc @Lorak-mmk

**Comment by Lorak-mmk** [comment]
Yes, lets do that so we can release 0.15.1

**Comment by wprzytula** [comment]
> This is API-breaking due to an error variant being removed. I suggest leaving the error variant for now under a TODO and removing it in the next major release. FYI @Lorak-mmk

Done.

**Review by Lorak-mmk — APPROVED** [review]

**Review by muzarski — APPROVED** [review]

**Comment by muzarski** [comment]
Actually - one more nit: aren't docstrings/comment related to this function outdated now?

**Comment by wprzytula** [comment]
> Actually - one more nit: aren't docstrings/comment related to this function outdated now?

Yees, good catch!

**Comment by wprzytula** [comment]
> Actually - one more nit: aren't docstrings/comment related to this function outdated now?

Fixed.

**Review by Lorak-mmk — DISMISSED** [review]

File: scylla-cql/src/frame/response/result.rs [hunk 3]
```diff
@@ -977,55 +977,24 @@ fn mk_col_spec_parse_error(
...
     for col_idx in 0..col_count {
-        let table_spec = deser_table_spec_for_col_spec(
-            buf,
-            global_table_spec_provided,
-            &mut known_table_spec,
-            col_idx,
-        )?;
+        let table_spec = match global_table_spec {
+            // If global table spec was provided, we simply clone it to each column spec.
+            Some(ref known_spec) if global_table_spec.is_some() => known_spec.clone(),
```
Comment (muzarski) [inline scylla-cql/src/frame/response/result.rs:991]: nit: we match against `global_table_spec` and have a match arm: `Some(_) if [...].is_some()`

Comment (muzarski) [reply scylla-cql/src/frame/response/result.rs:991]: I'm surprised clippy did not catch it

**Review by Lorak-mmk — APPROVED** [review]
